File: concurrency_paralellism/paralellism/sum.py

import multiprocessing

def sum_worker(chunk):
    return sum(chunk)

def parallel_sum(arr):
    num_processes = multiprocessing.cpu_count()
    chunk_size = max(1, len(arr) // num_processes)
    
    with multiprocessing.Pool(processes=num_processes) as pool:
        chunks = [arr[i:i+chunk_size] for i in range(0, len(arr), chunk_size)]
        results = pool.map(sum_worker, chunks)
    
    return sum(results)

if name == "main":
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(parallel_sum(arr))

File: concurrency_paralellism/paralellism/sum.rs
use std::thread;

fn parallel_sum(arr: &[i32]) -> i32 {
    const NUM_THREADS: usize = 4; // Number of threads for parallel processing
    let chunk_size = arr.len() / NUM_THREADS; // Divide the array into equal chunks
    
    let mut handles = vec![];
    let mut sums = vec![0; NUM_THREADS]; // Store partial sums for each chunk
    
    for (i, chunk) in arr.chunks(chunk_size).enumerate() {
        let handle = thread::spawn(move || {
            let sum = chunk.iter().sum(); // Calculate sum of the chunk
            sums[i] = sum; // Store the partial sum
        });
        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap(); // Wait for all threads to finish
    }

    // Calculate total sum by summing up partial sums
    let total_sum: i32 = sums.iter().sum();
    total_sum
}

fn main() {
    let arr = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let sum = parallel_sum(&arr);
    println!("{}", sum);
}

